src/tensorOps.hpp

```diff
@@ -22,126 +22,1990 @@
+ * @tparam T The type contained in the array.
+ * @tparam PROVIDED_M The size of the first dimension.
+ * @tparam PROVIDED_M The size of the second dimension.
+ */
+template< std::ptrdiff_t PROVIDED_M,
+          std::ptrdiff_t PROVIDED_N,
+          typename T,
+          std::ptrdiff_t INFERRED_M,
+          std::ptrdiff_t INFERRED_N >
+LVARRAY_HOST_DEVICE inline constexpr
+void checkSizes( T const ( & )[ INFERRED_M ][ INFERRED_N ] )
+{
+  static_assert( PROVIDED_M == INFERRED_M, "Expected the first dimension of size PROVIDED_M, got an array of size INFERRED_M." );
+  static_assert( PROVIDED_N == INFERRED_N, "Expected the second dimension of size PROVIDED_N, got an array of size INFERRED_N." );
+}
+
+/**
+ * @brief Verify at compile time that @tparam ARRAY is 1D and at runtime verify the size when USE_ARRAY_BOUNDS_CHECK.
+ * @tparam M The size expected size.
+ * @tparam ARRAY The type o @p array, should be an Array, ArrayView or ArraySlice.
+ * @param array The array to check.
+ */
+template< std::ptrdiff_t M, typename ARRAY >
+LVARRAY_HOST_DEVICE inline CONSTEXPR_WITHOUT_BOUNDS_CHECK
+void checkSizes( ARRAY const & array )
+{
+  static_assert( ARRAY::ndim == 1, "Must be a 1D array." );
+  #ifdef USE_ARRAY_BOUNDS_CHECK
+  LVARRAY_ERROR_IF_NE( array.size( 0 ), M );
+  #else
+  LVARRAY_UNUSED_VARIABLE( array );
+  #endif
+}
+
+/**
+ * @brief Verify at compile time that @tparam ARRAY is 2D and at runtime verify the sizes when USE_ARRAY_BOUNDS_CHECK.
+ * @tparam M The expected size of the first dimension.
+ * @tparam N The expected size of the second dimension.
+ * @tparam ARRAY The type o @p array, should be an Array, ArrayView or ArraySlice.
+ * @param array The array to check.
+ */
+template< std::ptrdiff_t M, std::ptrdiff_t N, typename ARRAY >
+LVARRAY_HOST_DEVICE inline CONSTEXPR_WITHOUT_BOUNDS_CHECK
+void checkSizes( ARRAY const & array )
+{
+  static_assert( ARRAY::ndim == 2, "Must be a 1D array." );
+#ifdef USE_ARRAY_BOUNDS_CHECK
+  LVARRAY_ERROR_IF_NE( array.size( 0 ), M );
+  LVARRAY_ERROR_IF_NE( array.size( 1 ), N );
+#else
+  LVARRAY_UNUSED_VARIABLE( array );
+#endif
+}
+
+/**
+ * @struct SquareMatrixOps
+ * @brief Performs operations on square matrices.
+ * @tparam M The size of the matrix.
+ * @note This class is intended to have its methods overwritten by template specializations
+ *   for specific matrix sizes, however generic methods could go here.
+ */
+template< int M >
+struct SquareMatrixOps
+{
+  static_assert( M > 0, "The size of the matrix must be greater than 0." );
+
+  /**
+   * @brief @return Return the determinant of the matrix @p matrix.
+   * @tparam MATRIX The type of @p matrix.
+   * @param matrix The matrix to get the determinant of.
+   * @note This is not implemented for generic sized matrices and calling this is a compilation error.
+   */
+  template< typename MATRIX >
+  LVARRAY_HOST_DEVICE CONSTEXPR_WITHOUT_BOUNDS_CHECK inline
+  static int determinant( MATRIX const & matrix )
```

Comment:
@castelletto1 I think it is kind of standard to just use `det()` for the function name....but I rather like spelling it out. Objections?